Source/Core/Core/CheatSearch.cpp

```diff
@@ -0,0 +1,377 @@
+  if (!matches_func)
+    return std::nullopt;
+
+  return NewSearchInternal(matches_func, value, type, GetValueSize(value));
+}
+
+static std::optional<std::vector<Cheats::SearchResult>>
+NextSearchInternal(const std::vector<Cheats::SearchResult>& previous_results,
+                   Cheats::CompareType comparison, const Cheats::SearchValue* value)
+{
+  if (!Memory::m_pRAM)
+    return std::nullopt;
+
+  std::vector<Cheats::SearchResult> results;
+  Core::RunAsCPUThread([&] {
+    for (const Cheats::SearchResult& previous_result : previous_results)
+    {
+      const u32 addr = previous_result.m_address;
+      if (!PowerPC::HostIsRAMAddress(addr, true))
+        continue;
+
+      const size_t value_size = GetValueSize(previous_result.m_value);
+      if (value_size == 0)
+        continue;
+
+      const Cheats::DataType type = Cheats::GetDataType(previous_result.m_value);
+      const auto matches_func = CreateMatchFunction(comparison, type);
+      if (!matches_func)
+        continue;
+
+      const auto current_value = ReadValueFromEmulatedMemory(addr, type, value_size);
+      if (matches_func(current_value, value ? *value : previous_result.m_value))
+      {
+        results.emplace_back();
+        results.back().m_address = addr;
+        results.back().m_value = current_value;
+      }
+    }
+  });
+
+  return results;
+}
+
+std::optional<std::vector<Cheats::SearchResult>>
+Cheats::NextSearch(const std::vector<Cheats::SearchResult>& previous_results,
+                   Cheats::CompareType comparison)
+{
+  return NextSearchInternal(previous_results, comparison, nullptr);
+}
+
+std::optional<std::vector<Cheats::SearchResult>>
+Cheats::NextSearch(const std::vector<Cheats::SearchResult>& previous_results,
+                   Cheats::CompareType comparison, const Cheats::SearchValue& value)
+{
+  return NextSearchInternal(previous_results, comparison, &value);
+}
+
+std::optional<std::vector<Cheats::SearchResult>>
+Cheats::UpdateValues(const std::vector<Cheats::SearchResult>& previous_results)
+{
+  if (!Memory::m_pRAM)
+    return std::nullopt;
+
+  std::vector<Cheats::SearchResult> results;
+  results.reserve(previous_results.size());
+  Core::RunAsCPUThread([&] {
+    for (const Cheats::SearchResult& previous_result : previous_results)
+    {
+      const u32 addr = previous_result.m_address;
+      if (!PowerPC::HostIsRAMAddress(addr, true))
+        continue;
+
+      const auto current_value = ReadValueFromEmulatedMemory(
+          addr, GetDataType(previous_result.m_value), GetValueSize(previous_result.m_value));
+      results.emplace_back();
+      results.back().m_address = addr;
+      results.back().m_value = current_value;
+    }
+  });
+
+  return results;
+}
+
+Cheats::DataType Cheats::GetDataType(const Cheats::SearchValue& value)
+{
+  // sanity checks that our enum matches with our std::variant
+  static_assert(
+      std::is_same<
+          u8, std::remove_cv<std::remove_reference<decltype(
```

Comment:
`std::is_same_v` and `std::remove_cv_t` can be used here to remove an instance of `::type` and `::value`